### PR TITLE
Fetch authentication parameters using the REST API

### PR DESCRIPTION
**Pull request by ThaminduR**
### Proposed changes in this pull request

- Add a wrapper to the request coming to the authentication endpoint.
- Use sessionDataKey/ SessionDataKeyConsent/ ErrorKey to get the necessary parameters from the REST API.
- Fix parameter filtering happening for federated identity providers.

Related Issue: https://github.com/wso2/product-is/issues/14878

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/client/model/AuthenticationRequestWrapper.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2022, WSO2 LLC. (http://www.wso2.com).
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
```
Comment (piyarathnalakmali) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/client/model/AuthenticationRequestWrapper.java:4]: Update the license header. Inc -> LLC

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/client/model/AuthenticationRequestWrapper.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        this.authParams = authParams;
+    }
+
+    /**
+     * Get the parameter from the map.
+     *
+     * @param name - Name of the parameter.
+     * @return
+     */
+    public String getParameter(String name) {
```
Comment (piyarathnalakmali) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/client/model/AuthenticationRequestWrapper.java:45]: Formatting error. Add a new line after the method signature.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+     * Build the API Path using the key and value.
+     *
+     * @param key   - SessionDataKey or SessionDataKeyConsent.
+     * @param value - Value of the key.
+     * @return - API Path.
+     */
+    private String buildAPIPath(String key, String value) {
+
+        try {
+            String authAPIURL = ServiceURLBuilder.create().addPath(DATA_API_PATH).build().getAbsoluteInternalURL();
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java:116]: If empty, use endpoint url from web.xml

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                    // Add 'sessionDataKeyConsent' to URL.
+                    authAPIURL += "data/OauthConsentKey/" + value;
+                    break;
+                case ERROR_KEY:
+                    authAPIURL += "data/AuthenticationError/" + value;
+                    break;
+            }
+            return authAPIURL;
+        } catch (URLBuilderException e) {
+            throw new RuntimeException(e);
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java:137]: Throw Auth Exception

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                return;
+            }
+        }
+        filterChain.doFilter(servletRequest, servletResponse);
+    }
+
+    /**
+     * Get HttpServletRequest with parameters retrieved from the API.
+     *
+     * @param servletRequest - Received HttpServletRequest.
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java:94]: remove -

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    private static final String SESSION_DATA_KEY_CONSENT = "sessionDataKeyConsent";
+    private static final String ERROR_KEY = "errorKey";
+
+    @Override
+    public void init(FilterConfig filterConfig) {
+        context = filterConfig.getServletContext();
+    }
+
+    /**
+     * @param servletRequest
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java:59]: provide explanations

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+     * @param servletResponse
+     * @param filterChain
+     * @throws IOException
+     * @throws ServletException
+     */
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain)
+            throws IOException, ServletException {
+        if (servletRequest instanceof HttpServletRequest) {
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java:68]: add new line

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.authentication.endpoint.util.AuthContextAPIClient;
+import org.wso2.carbon.identity.application.authentication.endpoint.util.client.model.AuthenticationRequestWrapper;
+import org.wso2.carbon.identity.core.ServiceURLBuilder;
+import org.wso2.carbon.identity.core.URLBuilderException;
+
+import java.io.IOException;
+import java.util.Map;
+import javax.servlet.Filter;
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java:31]: new line

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/client/model/AuthenticationRequestWrapper.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        try {
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/client/model/AuthenticationRequestWrapper.java:46]: replace with if block

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/client/model/AuthenticationRequestWrapper.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }
+
+    /**
+     * Get the parameter map.
+     *
+     * @return
+     */
+    public Map<String, Object> getParameterMap() {
+        return authParams;
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/client/model/AuthenticationRequestWrapper.java:59]: combine with super param map

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    /**
+     * Get HttpServletRequest with parameters retrieved from the API.
+     *
+     * @param servletRequest - Received HttpServletRequest.
+     * @param key            - Name of the key.
+     * @return - HttpServletRequest with parameters.
+     */
+    private HttpServletRequest getServletRequestWithParams(ServletRequest servletRequest, String key) {
+        String authAPIURL = buildAPIPath(key, servletRequest.getParameter(key));
+        String contextProperties = AuthContextAPIClient.getContextProperties(authAPIURL);
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java:100]: add correlation id to client

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java [hunk 12]
```diff
@@ -368,7 +368,8 @@ protected void unwrapResponse(CommonAuthResponseWrapper responseWrapper, String
 
         if (responseWrapper.isRedirect()) {
             String redirectURL;
-            if (context != null) {
+            if (context != null && (context.getExternalIdP() == null ||
+                    context.getExternalIdP().getIdPName().equals("LOCAL"))) {
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java:372]: "LOCAL".equal(context.getExternalIdP().getIdPName().equals)

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java [hunk 13]
```diff
@@ -612,8 +612,22 @@ public static void sendToRetryPage(HttpServletRequest request, HttpServletRespon
             URIBuilder uriBuilder = new URIBuilder(
                     ConfigurationFacade.getInstance().getAuthenticationEndpointRetryURL());
             if (status != null && statusMsg != null) {
-                uriBuilder.addParameter("status", status);
-                uriBuilder.addParameter("statusMsg", statusMsg);
+                if (context != null) {
+                    Map<String, String> failureData = new HashMap<>();
+                    failureData.put("status", status);
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:617]: Add constants

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java [hunk 14]
```diff
@@ -265,6 +265,8 @@ private void displayPrompt(AuthenticationContext context, HttpServletRequest req
                 tenantDomainQueryString = "tenantDomain=" + context.getTenantDomain();
                 promptPage = FrameworkUtils.appendQueryParamsStringToUrl(promptPage, tenantDomainQueryString);
             }
+            String sessionDataKeyQueryString = "sessionDataKey=" + context.getContextIdentifier();
+            promptPage = FrameworkUtils.appendQueryParamsStringToUrl(promptPage, sessionDataKeyQueryString);
             String redirectUrl = FrameworkUtils.appendQueryParamsStringToUrl(promptPage, "templateId=" +
                     URLEncoder.encode(promptNode.getTemplateId(), StandardCharsets.UTF_8.name()) + "&promptId=" +
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java:271]: replace promptId using sessionDataKey

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java [hunk 15]
```diff
@@ -322,7 +322,7 @@ public void handle(HttpServletRequest request, HttpServletResponse response) thr
                     FrameworkUtils.getPASTRCookieName(context.getContextIdentifier()));
             publishAuthenticationFailure(request, context, context.getSequenceConfig().getAuthenticatedUser(),
                     e.getErrorCode());
-            FrameworkUtils.sendToRetryPage(request, responseWrapper, errorWrapper.getStatus(),
+            FrameworkUtils.sendToRetryPage(request, responseWrapper, context, errorWrapper.getStatus(),
                     errorWrapper.getStatusMsg());
         } catch (Throwable e) {
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java:327]: Change to exception

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java (hunk 3 above)
Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java:116]: - Tenant Resolve URL
- Use template for server port

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java (hunk 13 above)
Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/FrameworkUtils.java:617]: Updated existing `STATUS` and `STATUS_MSG` values in `FrameworkConstants` and used those here.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java (hunk 11 above)
Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java:114]: Tracked via https://github.com/wso2/product-is/issues/14841

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java (hunk 14 above)
Comment (ThaminduR) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/GraphBasedSequenceHandler.java:271]: Tracked via https://github.com/wso2/product-is/issues/14842

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/client/exception/AuthenticationEndpointException.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.application.authentication.endpoint.util.client.exception;
+
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/client/exception/AuthenticationEndpointException.java:20]: class level comment

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+    private static final String SESSION_DATA_KEY_CONSENT = "sessionDataKeyConsent";
+    private static final String ERROR_KEY = "errorKey";
+
+    @Override
+    public void init(FilterConfig filterConfig) {
+        context = filterConfig.getServletContext();
+    }
+
+    /**
+     * @param servletRequest Servlet Request
```
Comment (ThaminduR) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java:65]: fix formatting

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/client/model/AuthenticationRequestWrapper.java [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+            return (authParams.get(name) != null) ? (String) authParams.get(name) : super.getParameter(name);
+        } else {
+            return super.getParameter(name);
+        }
+    }
+
+    /**
+     * Get the parameter map.
+     *
+     * @return
```
Comment (sahandilshan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/client/model/AuthenticationRequestWrapper.java:74]: Incompleted method comment. Fix in the other places as well

**Review by senthalan — APPROVED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/AuthenticationEndpointUtil.java [hunk 19]
```diff
@@ -435,4 +444,26 @@ private static void setAuthorizationHeader(HttpRequestBase httpMethod) {
 
         httpMethod.addHeader(HTTPConstants.HEADER_AUTHORIZATION, CLIENT + authHeader);
     }
+
+    /**
+     * Resolve "${tenantDomain}" in the URL.
+     *
+     * @param   url URL to be tenant resolved
```
Comment (sandushi) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/AuthenticationEndpointUtil.java:451]: Shall we add a full stop after the each line of the comment to follow the same as other comments. Check the other places as well

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+    public void init(FilterConfig filterConfig) {
+        context = filterConfig.getServletContext();
+    }
+
+    /**
+     * Retrieve the parameters for the received servlet request and forward the request wrapped with the parameters.
+     *
+     * @param servletRequest    Servlet Request
+     * @param servletResponse   Servlet Response
+     * @param filterChain       Filter Chain
```
Comment (sandushi) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint.util/src/main/java/org/wso2/carbon/identity/application/authentication/endpoint/util/filter/AuthParameterFilter.java:69]: Shall we add full stop at the end of each comment

**Pull request merged**